### PR TITLE
feat(builtins): implement regex

### LINKED ISSUE #137: builtins: implement regex
### What is the underlying problem you're trying to solve?

Implement [`regex`](https://www.openpolicyagent.org/docs/policy-reference/builtins/regex) builtins from the OPA specification: 

- [ ] [regex.find_all_string_submatch_n](https://www.openpolicyagent.org/docs/policy-reference/builtins/regex#builtin-regex-regexfind_all_string_submatch_n)                             
- [ ] [regex.find_n](https://www.openpolicyagent.org/docs/policy-reference/builtins/regex#builtin-regex-regexfind_n)
- [ ] [regex.globs_match](https://www.openpolicyagent.org/docs/policy-reference/builtins/regex#builtin-regex-regexglobs_match)
- [ ] [regex.is_vald](https://www.openpolicyagent.org/docs/policy-reference/builtins/regex#builtin-regex-regexis_valid)
- [ ] [regex.match](https://www.openpolicyagent.org/docs/policy-reference/builtins/regex#builtin-regex-regexmatch)
- [ ] [regex.replace](https://www.openpolicyagent.org/docs/policy-reference/builtins/regex#builtin-regex-regexreplace)
- [ ] [regex.split](https://www.openpolicyagent.org/docs/policy-reference/builtins/regex#builtin-regex-regexsplit)
- [ ] [regex.template_match](https://www.openpolicyagent.org/docs/policy-reference/builtins/regex#builtin-regex-regextemplate_match)

### Expected Outcome

- `regex` builtins can be used in policy evaluations
- All *regex* tests pass:
  - https://github.com/open-policy-agent/opa/tree/main/v1/test/cases/testdata/v1/regexfind
  - https://github.com/open-policy-agent/opa/tree/main/v1/test/cases/testdata/v1/regexfindallstringsubmatch
  - https://github.com/open-policy-agent/opa/tree/main/v1/test/cases/testdata/v1/regexisvalid
  - https://github.com/open-policy-agent/opa/tree/main/v1/test/cases/testdata/v1/regexmatch
  - https://github.com/open-policy-agent/opa/tree/main/v1/test/cases/testdata/v1/regexsplit
  - https://github.com/open-policy-agent/opa/tree/main/v1/test/cases/testdata/v1/regexreplace
  - https://github.com/open-policy-agent/opa/tree/main/v1/test/cases/testdata/v1/regexmatchtemplate

### PR DESCRIPTION
**Pull request by DFrenkel**
### What code changed, and why?

Implement all 7 `regex.*` builtins from the OPA specification, completing regex support for swift-opa.

Implementation uses Swift's native `Regex<AnyRegexOutput>` instead of `NSRegularExpression`, avoiding NSString bridging overhead. `RegexCache` matches Go OPA's package-level `regexpCache` (100-entry cap, random eviction, `NSLock`). Templates and raw patterns use separate cache dictionaries to avoid key collisions.

`regex.replace` includes a manual `$N` capture group template expander (Swift `Regex` has no equivalent to Go's `ReplaceAllString` template API) with two-pass string assembly to avoid value-type reallocations.

`regex.template_match` is ported from Go OPA's `compileRegexTemplate` - it parses balanced delimiters, escapes literal segments, and compiles the result.

### Definition of done

- [x] All 7 regex builtins implemented and registered
- [x] All 27 compliance test cases pass
- [x] Unit tests cover compliance cases, edge cases, error cases, and type/arg validation
- [x] `RegexCache` documents upgrade path for threading (`DispatchQueue`, `Mutex`, `pthread_rwlock_t`)
- [x] `capabilities.json` updated

### How to test

```
make test
OPA_COMPLIANCE_TESTS="regex" make test-compliance
```

### Related Resources

Closes #137

### CODE REVIEW

File: Sources/Rego/Builtins/Regex.swift [hunk 1]
```diff
@@ -0,0 +1,450 @@
...
+///
+/// - `DispatchQueue(.concurrent)` with barrier writes for true concurrent reads.
+/// - `Mutex` (Swift `Synchronization` module) when the deployment target
+///   reaches macOS 15+ / iOS 18+.
+/// - `pthread_rwlock_t` for shared-reader/exclusive-writer semantics, with
+///   `#if os(Windows)` conditional for `SRWLOCK`.
+final class RegexCache: @unchecked Sendable {
+    private var patterns: [String: Regex<AnyRegexOutput>] = [:]
+    private var templates: [String: Regex<AnyRegexOutput>] = [:]
+    private let lock = NSLock()
```
Comment (FranzBusch) [inline Sources/Rego/Builtins/Regex.swift:17]: Can we use Mutex here?

File: Sources/Rego/Builtins/Regex.swift [hunk 2]
```diff
@@ -0,0 +1,450 @@
...
+/// if the underlying `Dictionary` structure weren't being mutated by inserts.
+/// The lock is held only for dictionary operations (~100ns), so contention is
+/// negligible in practice. Upgrade path if this changes:
+///
+/// - `DispatchQueue(.concurrent)` with barrier writes for true concurrent reads.
+/// - `Mutex` (Swift `Synchronization` module) when the deployment target
+///   reaches macOS 15+ / iOS 18+.
+/// - `pthread_rwlock_t` for shared-reader/exclusive-writer semantics, with
+///   `#if os(Windows)` conditional for `SRWLOCK`.
+final class RegexCache: @unchecked Sendable {
```
Comment (FranzBusch) [inline Sources/Rego/Builtins/Regex.swift:14]: If we use Mutex this shouldn't need an `@unchecked Sendable`

File: Sources/Rego/Builtins/Regex.swift (hunk 1 above)
Comment (philipaconrad) [reply Sources/Rego/Builtins/Regex.swift:17]: `Mutex` is only [available on macOS 15+ targets](https://developer.apple.com/documentation/synchronization/mutex), and we're on macOS 13 at present. That might explain the use of `NSLock`, instead of `Mutex`?

Comment (DFrenkel) [reply Sources/Rego/Builtins/Regex.swift:17]: > `Mutex` is only [available on macOS 15+ targets](https://developer.apple.com/documentation/synchronization/mutex), and we're on macOS 13 at present. That might explain the use of `NSLock`, instead of `Mutex`?

Correct, that was the intent @philipaconrad

File: Sources/Rego/Builtins/Regex.swift (hunk 2 above)
Comment (DFrenkel) [reply Sources/Rego/Builtins/Regex.swift:14]: See discussion with @philipaconrad about target versions. Is that acceptable to leave as-is for now?

File: README.md [hunk 3]
```diff
@@ -197,6 +197,46 @@ var engine = OPA.Engine(
 let preparedQuery = try await engine.prepareForEvaluation(query: "<some_query>")
 ```
 
+## Regex syntax — RE2 only
```
Comment (philipaconrad) [inline README.md]: I think for behavioral differences like this, we might do well to put it under a general section header like "`## Known Differences from OPA`", and then use [`<details>...</details>` blocks](https://docs.github.com/en/get-started/writing-on-github/working-with-advanced-formatting/organizing-information-with-collapsed-sections)[^1] to hide each section/group of differing behaviors.

In this PR, it will be regex differences, in a later one, we might document numerics differences, and so on. 🤔 I can expand on this, or start the differences section in a separate PR if you'd find that useful.

[^1]: One of my favorite uses for collapsed sections blocks is for hiding detailed breakdowns of benchmark results after the summary, so that most reviewers can skip the wall of text, and those who want the details can still see them. (Examples: [opa #8664](https://github.com/open-policy-agent/opa/pull/8664), [opa #8693](https://github.com/open-policy-agent/opa/pull/8693))

**Comment by philipaconrad** [comment]
ℹ️ @DFrenkel if you see spurious errors in CI for older targets (Swift <6.2), try rebasing your PR to catch up with `main`-- I just merged #175, which should eliminate the random toolchain errors.

File: CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,22 @@ project adheres to [Semantic Versioning](http://semver.org/).
 
 ## Unreleased
 
+### `regex` builtins (RE2-only contract)
```
Comment (philipaconrad) [inline CHANGELOG.md]:
```suggestion
### `regex` builtins
```
[nit]: Let's just call out that the builtins are here, and then explain the "RE2 is the only supported syntax" bits in the description below".

File: CHANGELOG.md [hunk 5]
```diff
@@ -5,6 +5,22 @@ project adheres to [Semantic Versioning](http://semver.org/).
...
+- `regex.match`
+- `regex.replace`
+- `regex.split`
+- `regex.find_n`
+- `regex.find_all_string_submatch_n`
+- `regex.template_match`
+
+> **Heads up — RE2 is the long-term contract.** Only Go [RE2](https://github.com/google/re2/wiki/Syntax) syntax is supported, and only RE2 syntax is *guaranteed* to keep working in future releases. Patterns that happen to work today purely because the current implementation is built on `NSRegularExpression` (ICU) — for example backreferences, lookaround, atomic groups, or ICU-style named groups — are not part of the contract and are not guaranteed to keep working. **Do not rely on non-RE2 features.**
+
+To enforce this, every pattern is validated against an RE2 compatibility scanner up-front. Patterns using backreferences (`\1`–`\9`), lookahead (`(?=...)`, `(?!...)`), lookbehind (`(?<=...)`, `(?<!...)`), or atomic groups (`(?>...)`) are rejected at compile time — `regex.is_valid` returns `false`, and the other builtins throw. The scanner will be tightened over time. See the README's "Regex syntax — RE2 only" section for the full contract.
```
Comment (philipaconrad) [inline CHANGELOG.md]:
```suggestion
To enforce this, every pattern is validated against an RE2 compatibility scanner up-front. Patterns using backreferences (`\1`–`\9`), lookahead (`(?=...)`, `(?!...)`), lookbehind (`(?<=...)`, `(?<!...)`), or atomic groups (`(?>...)`) are rejected at compile time: `regex.is_valid` returns `false`, and the other builtins throw. The scanner will be tightened over time. See the README's "Regex syntax support" section for more details.
```
[nit]: Small tweaks + corrected section ref.

File: CHANGELOG.md [hunk 6]
```diff
@@ -5,6 +5,22 @@ project adheres to [Semantic Versioning](http://semver.org/).
...
+
+- `regex.is_valid`
+- `regex.match`
+- `regex.replace`
+- `regex.split`
+- `regex.find_n`
+- `regex.find_all_string_submatch_n`
+- `regex.template_match`
+
+> **Heads up — RE2 is the long-term contract.** Only Go [RE2](https://github.com/google/re2/wiki/Syntax) syntax is supported, and only RE2 syntax is *guaranteed* to keep working in future releases. Patterns that happen to work today purely because the current implementation is built on `NSRegularExpression` (ICU) — for example backreferences, lookaround, atomic groups, or ICU-style named groups — are not part of the contract and are not guaranteed to keep working. **Do not rely on non-RE2 features.**
```
Comment (philipaconrad) [inline CHANGELOG.md]:
```suggestion
> [!WARNING]
> Swift OPA's `regex.*` builtins are only planned to support [RE2](https://github.com/google/re2/wiki/Syntax) syntax in the long term. Any regex syntax outside of RE2 syntax is unsupported, and may be broken in a future release.
```
[suggestion]: We can copy the warning block over from the README wholesale, and Github *should* render the `WARNING` block correctly in the release notes later on. 🤔

File: README.md (hunk 3 above)
Comment (DFrenkel) [reply README.md]: Fixed README and CHANGELOG

**Review by philipaconrad — DISMISSED** [review]
Thank you @DFrenkel for your work on this PR!

I know it took a while, but I'm excited to have regex support landing at long last. 😄 This will be a fun set of builtins to have going out in the next release!

File: Sources/Rego/Builtins/Regex.swift (hunk 2 above)
Comment (philipaconrad) [reply Sources/Rego/Builtins/Regex.swift:14]: @DFrenkel I think it's fine to leave as-is for now. If/when we decide to bump the minimum target version, we should refactor to use Mutexes where it makes sense to do so.

File: CHANGELOG.md (hunk 6 above)
Comment (DFrenkel) [reply CHANGELOG.md]: Fixed

File: CHANGELOG.md (hunk 5 above)
Comment (DFrenkel) [reply CHANGELOG.md]: Fixed, thank you!

File: CHANGELOG.md (hunk 4 above)
Comment (DFrenkel) [reply CHANGELOG.md]: Fixed

**Review by philipaconrad — APPROVED** [review]
